javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -0,0 +1,153 @@
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(leading), offset, newSize, depthShift);
+        } else {
+            Object[] array = this.array;
+            int depthShift = this.depthShift, offset = this.offset;
+            if (isFullLeft()) {
+                array = copyUpdate(Arrays.empty(), branchingFactor() - 1, array);
+                depthShift += BRANCHING_BASE;
+                offset = treeSize(branchingFactor() - 1, depthShift);
+            }
+
+            offset -= 1;
+            array = modifyLeaf(array, depthShift, offset, NODE_MODIFIER, LEAF_MODIFIER.apply(leading));
+            return new BitMappedTrie<>(array, offset, newSize, depthShift);
+        }
+    }
+    private boolean isFullLeft() { return offset == 0; }
+
+    BitMappedTrie<T> appendLeaf(T trailing) {
+        final int newSize = length() + 1;
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(trailing), offset, newSize, depthShift);
+        } else {
+            Object[] array = this.array;
+            int depthShift = this.depthShift;
+            if (isFullRight(newSize)) {
+                array = asArray(array);
+                depthShift += BRANCHING_BASE;
+            }
+
+            array = modifyLeaf(array, depthShift, offset + length(), NODE_MODIFIER, LEAF_MODIFIER.apply(trailing));
+            return new BitMappedTrie<>(array, offset, newSize, depthShift);
+        }
+    }
+    private boolean isFullRight(int newSize) { return (offset + newSize) > treeSize(branchingFactor(), depthShift); }
+
+    private static final NodeModifier NODE_MODIFIER = (o, i) -> copy(o, i + 1);
+    private static final Function<Object, NodeModifier> LEAF_MODIFIER = element -> ((o, i) -> copyUpdate(o, i, element));
+
+    BitMappedTrie<T> update(int index, T element) {
+        final Object[] root = modifyLeaf(array, depthShift, offset + index, NODE_MODIFIER, LEAF_MODIFIER.apply(element));
+        return new BitMappedTrie<>(root, offset, length(), depthShift);
+    }
+
+    BitMappedTrie<T> drop(int n) {
+        if (n <= 0) {
+            return this;
+        } else if (n >= length()) {
+            return empty();
+        } else {
+            final NodeModifier drop = Arrays::copyDrop;
+            final Object[] root = modifyLeaf(array, depthShift, offset + n, drop, drop);
+            return new BitMappedTrie<>(root, offset + n, length() - n, depthShift);
+        }
+    }
+
+    BitMappedTrie<T> take(int n) {
+        if (n >= length()) {
+            return this;
+        } else if (n <= 0) {
+            return empty();
+        } else {
+            final NodeModifier take = Arrays::copyTake;
+            final Object[] root = modifyLeaf(array, depthShift, (offset + n) - 1, take, take);
+            return new BitMappedTrie<>(root, offset, n, depthShift);
+        }
+    }
+
+    private static Object[] modifyLeaf(Object[] root, int depthShift, int index, NodeModifier node, NodeModifier leaf) {
```

Comment:
+1 to this method: it's a really great idea to have a single method to traverse the tree, with the behavior for modifying the nodes and the leaf being parametric.